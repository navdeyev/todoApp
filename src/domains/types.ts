import {RouterState} from 'react-router-redux';

import {ITheme} from 'components/Styled/themes';
import {MediaType} from 'components/Styled/utils';
import {LoadingStates} from 'domains/loadingStates';
import {ITodo} from 'domains/todos/todosTypes';
import {IWindowDimensions} from 'domains/window/windowTypes';

export interface ITodosState {
    loadingState: LoadingStates,
    todos: ITodo[],
    selectedTodoId: string
}

export interface IWindowState {
    windowDimensions: IWindowDimensions,
    media: MediaType
}

export interface IThemeState {
    theme: ITheme
}

export interface IAppState {
    routing: RouterState,
    todosState: ITodosState,
    windowState: IWindowState,
    themeState: IThemeState
}
