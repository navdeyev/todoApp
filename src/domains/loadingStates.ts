export enum LoadingStates {
    NOT_STARTED = 'NOT_STARTED',
    LOADING = 'LOADING',
    COMPLETE = 'COMPLETE',
}