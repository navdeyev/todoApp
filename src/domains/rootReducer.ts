import {routerReducer} from 'react-router-redux';
import {combineReducers} from 'redux';

import {IAppState} from './types';

import todosReducer from 'domains/todos/todosReducer';
import windowReducer from 'domains/window/windowReducer';

const rootReducer = combineReducers<IAppState>({
    routing: routerReducer,
    todosState: todosReducer,
    windowState: windowReducer
});

export default rootReducer;