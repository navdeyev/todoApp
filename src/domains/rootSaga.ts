import {all, fork} from 'redux-saga/effects';

import todosSaga from './todos/todosSaga';

import {IServiceMap} from 'service/services';

function* rootSaga(serviceMap: IServiceMap) {
    yield all([
        fork(todosSaga, serviceMap)
    ]);
}

export default rootSaga;