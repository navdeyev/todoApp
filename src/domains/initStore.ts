import {applyMiddleware, createStore, Store} from 'redux';
import {composeWithDevTools} from 'redux-devtools-extension'
import thunk from 'redux-thunk';

import rootReducer from './rootReducer';
import {IAppState} from './types';

const initStore = (): Store<IAppState> => {
    const middlewares = [
        thunk
    ];

    const enhancer = process.env.NODE_ENV === 'development'
        ? composeWithDevTools(applyMiddleware(...middlewares))
        : applyMiddleware(...middlewares);

    return createStore<IAppState>(rootReducer, enhancer);
};

export default initStore;