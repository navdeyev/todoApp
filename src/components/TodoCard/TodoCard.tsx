import * as React from 'react';
import {connect} from 'react-redux';
import {ActionCreator, AnyAction} from 'redux';
import styled from 'styled-components';

import * as todosActions from 'domains/todos/todosActions';
import {IAppState, ITodo} from 'domains/types';

import {colors} from 'components/Styled/colors';
import {Heading2, Paragraph} from 'components/Styled/Styled';
import {rem} from 'components/Styled/utils';
import TodoStepList from 'components/TodoStepsList/TodoStepList';

const Wrapper = styled.div`
    width: 100%;
    box-sizing: border-box;
    border: 2px solid ${ colors.GREY };
    border-radius: ${ rem(6) };
    margin-bottom: ${ rem(16) };
    padding: ${ rem(20) } ${ rem(16) };
`;
Wrapper.displayName = 'Wrapper';

const HeaderWrapper = styled.div`
    display: flex;
    flex-direction: row;
`;
HeaderWrapper.displayName = 'HeaderWrapper';

const CardHeader = Heading2.extend`flex: 5 1 0px;`;
CardHeader.displayName = 'CardHeader';

const Status = styled.div`
    flex: 1 1 0px;
    display: flex;
    padding: 0 ${ rem(5) }
    align-items: center;
    justify-content: center;
`;
Status.displayName = 'Status';

const Goal = Paragraph.extend`
    cursor: pointer;
    padding-top: ${ rem(8) };
    text-decoration: underline;

    &:hover {
      text-decoration: none;
    }
`;
Goal.displayName = 'Goal';

export interface IProps {
    todo: ITodo;
    selectedTodoId: string;
    selectTodo: ActionCreator<AnyAction>
}

export const TodoCard: React.SFC<IProps> = (props) => {
    const {todo, selectedTodoId, selectTodo} = props;

    const createClickHandler = (todoId: string) => () => {
        selectTodo(todoId);
    };

    return (
        <Wrapper>
            <HeaderWrapper>
                <CardHeader>{todo.title}</CardHeader>
                <Status>{todo.status}</Status>
            </HeaderWrapper>
            <Goal onClick={createClickHandler(todo.id)}>{todo.goal}</Goal>
            {selectedTodoId === todo.id && <TodoStepList steps={todo.steps}/>}
        </Wrapper>
    );
};

const mapStateToProps = (state: IAppState) => {
    return {
        selectedTodoId: state.todosState.selectedTodoId
    };
};

const mapDispatchToProps = {
    selectTodo: todosActions.selectTodo
};
export default connect(mapStateToProps, mapDispatchToProps)(TodoCard);