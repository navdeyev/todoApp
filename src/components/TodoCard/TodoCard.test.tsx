import {shallow} from 'enzyme';
import * as React from 'react';

import {TodoStatus} from 'domains/types';

import TodoCard, {IProps} from './TodoCard';

describe('TodoCard', () => {

    let defaultProps: IProps;
    beforeEach(() => {
        defaultProps = {
            todo: {
                goal: 'Goal description',
                id: 'some-id',
                status: TodoStatus.COMPLETE,
                steps: [],
                title: 'Todo title'
            }
        }
    });

    it('renders a component', () => {
        const render = shallow(<TodoCard {...defaultProps} />);
        expect(render.getElement()).toMatchSnapshot();
    });

});