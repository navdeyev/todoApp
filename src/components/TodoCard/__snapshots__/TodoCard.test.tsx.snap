// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoCard renders a component with expanded TodoStepList 1`] = `
<Wrapper>
  <HeaderWrapper>
    <CardHeader>
      Todo title
    </CardHeader>
    <Status>
      COMPLETE
    </Status>
  </HeaderWrapper>
  <Goal
    onClick={[Function]}
  >
    Goal description
  </Goal>
  <TodoStepList
    steps={Array []}
  />
</Wrapper>
`;

exports[`TodoCard renders a component with non-expanded TodoStepList 1`] = `
<Wrapper>
  <HeaderWrapper>
    <CardHeader>
      Todo title
    </CardHeader>
    <Status>
      COMPLETE
    </Status>
  </HeaderWrapper>
  <Goal
    onClick={[Function]}
  >
    Goal description
  </Goal>
</Wrapper>
`;
